### PR TITLE
Fix allow_none to work correctly on BaseInstance objects

### LINKED ISSUE #495: allow_none given to Date and Time are not respected
`allow_none` value given to `Date` and `Time` are not respected.

To reproduce:
```
# tmp.py
import datetime
from traits.api import HasTraits, Date, Time, Instance


class Foo(HasTraits):
    date = Date(allow_none=False)
    time = Time(allow_none=False)
    instance_date = Instance(datetime.date, allow_none=False)
    instance_time = Instance(datetime.time, allow_none=False)


foo = Foo()
```

Expected these two lines to raise TraitError, but they pass:
```
In [1]: %run tmp.py

In [2]: foo.date = None

In [3]: foo.time = None

```

One can workaround using `Instance`:
```
In [4]: foo.instance_date = None
---------------------------------------------------------------------------
TraitError                                Traceback (most recent call last)
<ipython-input-4-fa663306bb41> in <module>()
----> 1 foo.instance_date = None

/Users/kchoi/.edm/envs/resist-py36/lib/python3.6/site-packages/traits/trait_handlers.py in error(self, object, name, value)
    234         """
    235         raise TraitError(
--> 236             object, name, self.full_info(object, name, value), value
    237         )
    238 

TraitError: The 'instance_date' trait of a Foo instance must be a date, but a value of None <class 'NoneType'> was specified.

In [5]: foo.instance_time = None
---------------------------------------------------------------------------
TraitError                                Traceback (most recent call last)
<ipython-input-5-adf6a97a17be> in <module>()
----> 1 foo.instance_time = None

/Users/kchoi/.edm/envs/resist-py36/lib/python3.6/site-packages/traits/trait_handlers.py in error(self, object, name, value)
    234         """
    235         raise TraitError(
--> 236             object, name, self.full_info(object, name, value), value
    237         )
    238 

TraitError: The 'instance_time' trait of a Foo instance must be a time, but a value of None <class 'NoneType'> was specified.


```

### PR DESCRIPTION
**Pull request by mdickinson**
Currently instances of `BaseInstance` don't clone themselves correctly when called with a new version of `allow_none`. This means that `Time(allow_none=False)` and `Datetime(allow_none=False)` don't work as expected:

Example (on the master branch):

```python
>>> from traits.api import *
>>> Slice = BaseInstance(slice)
>>> SliceNotNone = Slice(allow_none=False)
>>> Slice.validate(None, "dummy", None)  # accepts None, as expected
>>> SliceNotNone.validate(None, "dummy", None)  # accepts None, but shouldn't
```

This PR fixes the issue by extending the base class `clone` method to add special handling of `allow_none`. With the PR, the above example becomes:

```python
>>> from traits.api import *
>>> Slice = BaseInstance(slice)
>>> SliceNotNone = Slice(allow_none=False)
>>> Slice.validate(None, "dummy", None)  # accepts None, as expected
>>> SliceNotNone.validate(None, "dummy", None)  # rejects None, as expected  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 3361, in validate
    self.validate_failed(object, name, value)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/trait_types.py", line 3211, in validate_failed
    self.error(object, name, value)
  File "/Users/mdickinson/Enthought/ETS/traits/traits/base_trait_handler.py", line 75, in error
    object, name, self.full_info(object, name, value), value
traits.trait_errors.TraitError: The 'dummy' trait must be a slice, but a value of None <class 'NoneType'> was specified.
```

We also take the opportunity to add some basic tests for the `Datetime` and `Time` trait types.

Fixes #495

**Checklist**
- [x] Tests
- ~[ ] Update API reference (`docs/source/traits_api_reference`)~ N/A
- ~[ ] Update User manual (`docs/source/traits_user_manual`)~ N/A
- ~[ ] Update type annotation hints in `traits-stubs`~ N/A

### CODE REVIEW

**Comment by mdickinson** [comment]
Note: while this _is_ a bugfix, it's a long-standing bug that hasn't caused major issues, so it doesn't seem worth backporting to any bugfix release on the 6.2.x branch.

File: traits/trait_type.py [hunk 1]
```diff
@@ -281,6 +281,11 @@ def clone(self, default_value=NoDefaultSpecified, **metadata):
             A dictionary of metadata names and corresponding values as
             arbitrary keyword arguments.
 
+        Returns
```
Comment (mdickinson) [inline traits/trait_type.py:284]: Drive-by documentation fix; no associated change in behaviour.

File: traits/trait_types.py [hunk 2]
```diff
@@ -3424,6 +3428,14 @@ def get_default_value(self):
 
         return (dvt, dv)
 
+    def clone(self, default_value=NoDefaultSpecified, **metadata):
+        """ pass """
```
Comment (rahulporuri) [inline traits/trait_types.py:3432]: I think it'll be useful to have a better docstring than this or just no docstring at all.

**Review by rahulporuri — APPROVED** [review]
LGTM with one question - I'm not sure why this fix only works on `BaseInstance` objects and not on `BaseInstance` subclasses? Won't the subclasses inherit the behavior via the inherited `clone` method?

File: traits/trait_types.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_types.py:3432]: Aargh! That was never intended to end up in the finished PR. Will fix.

**Comment by mdickinson** [comment]
> I'm not sure why this fix only works on `BaseInstance` objects and not on `BaseInstance` subclasses?

It works on both. I'll fix the title to be less misleading. (The point of the title is that the usual pattern is `some_trait_class(allow_none=True)`, where `allow_none` is a parameter to the initializer of the `TraitType` *subclass*, but the bug we're fixing is for the case of `trait_instance(allow_none=True)`, where `trait_instance` is already an *instance* of `BaseInstance`.

**Comment by mdickinson** [comment]
To be clear: before this PR

- If `SomeTraitType` is a subclass of `BaseInstance`, then `foo = SomeTraitType(allow_none=False)` already works as expected
- If `SomeThing` is an instance of `BaseInstance` (as e.g., `Datetime` and `Time` are), then `foo = SomeThing(allow_none=False)` does not work as intended - `None` is still allowed

This PR fixes the second part above.

File: traits/trait_types.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_types.py]: Fixed
